reposerver/repository/repository.go

```diff
@@ -1087,17 +1087,25 @@ func findPlugin(plugins []*v1alpha1.ConfigManagementPlugin, name string) *v1alph
 	return nil
 }
 
-func runConfigManagementPlugin(appPath string, envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds git.Creds) ([]*unstructured.Unstructured, error) {
-	concurrencyAllowed := isConcurrencyAllowed(appPath)
-	if !concurrencyAllowed {
-		manifestGenerateLock.Lock(appPath)
-		defer manifestGenerateLock.Unlock(appPath)
-	}
-
+func runConfigManagementPlugin(appPath, repoRoot string, envVars *v1alpha1.Env, q *apiclient.ManifestRequest, creds git.Creds) ([]*unstructured.Unstructured, error) {
 	plugin := findPlugin(q.Plugins, q.ApplicationSource.Plugin.Name)
 	if plugin == nil {
-		return nil, fmt.Errorf("Config management plugin with name '%s' is not supported.", q.ApplicationSource.Plugin.Name)
+		return nil, fmt.Errorf("config management plugin with name '%s' is not supported", q.ApplicationSource.Plugin.Name)
 	}
+
+	// Plugins can request to lock the complete repository when they need to
+	// use git client operations.
+	if plugin.LockRepo {
+		manifestGenerateLock.Lock(repoRoot)
+		defer manifestGenerateLock.Unlock(repoRoot)
+	} else {
+		concurrencyAllowed := isConcurrencyAllowed(appPath)
```

Comment:
The `isConcurrencyAllowed` method already returns false for plugins. Unless I'm missing something plugins already lock repo. We probably need opposite setting  `allowConcurrency` for plugins

Reply:
As far as I could tell, it locks on the `appPath`, but we need to lock at the repository root to prevent concurrent `git` operations (i.e. a plugin calling `git` in the same repo, but on different paths).

Reply:
That [`isConcurrencyAllowed`](https://github.com/argoproj/argo-cd/blob/40629e5715c3726c0bf97902eab67501e22c79b2/reposerver/repository/repository.go#L511) method only checks for `.argocd-allow-concurrency`, which could exist for plugins,  unless I am missing something? I had assumed that the general point of the logic here was about forcing this method to lock even if that file was specified (if the plugin is configured with lockrepo).